File: 2023/rust/day01/src/bin/part1.rs
fn main() {
    let input = include_str!("./input.txt");
    let output = solve(input);
    dbg!(output);
}

fn solve(input: &str) -> String {
    input
        .lines()
        .map(|line| {
            line
                .chars()
                .filter_map(|character| {
                    character.to_digit(10)
                })
                .collect::<Vec<u32>>()
        })
        .map(|digits| (digits.first().unwrap() * 10) + digits.last().unwrap())
        .sum::<u32>()
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_solves_the_example() {
        let input = "1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet";
        assert_eq!(solve(input), "142");
    }
    
    #[test]
    fn it_solves_the_puzzle() {
        let input = include_str!("./input.txt");
        assert_eq!(solve(input), "53334");
    }
}

File: 2023/haskell/day01/part1.hs
module Main where

import Data.FileEmbed (embedStringFile, makeRelativeToProject)
import Data.Char

main :: IO ()
main = do
  print . solve $ input
  where input = $(makeRelativeToProject "input.txt" >>= embedStringFile)

solve :: String -> String
solve = show . sum . map calibrateValue . lines

calibrateValue :: String -> Int
calibrateValue value =
  let digits = filter isDigit value
      firstDigit = digitToInt $ head digits
      lastDigit = digitToInt $ last digits
   in firstDigit * 10 + lastDigit